obo_parser.py

```diff
@@ -0,0 +1,388 @@
+
+        for record in category_subtree:
+            if add_category_id_column:
+                record['category_id'] = category_id
+            if add_category_name_column:
+                record['category_name'] = category_name
+
+
+def _open_input_stream(path):
+    """Returns an open stream for iterating over lines in the given path.
+
+    Args:
+        path (str): url or local file path
+    Return:
+        iter: iterator over file handle
+    """
+    if not isinstance(path, basestring):
+        raise ValueError("Unexpected path type: %s" % type(path))
+
+    is_url = path.startswith("http")
+    if is_url:
+        line_iterator = contextlib.closing(urllib.urlopen(path))
+    else:
+        if not os.path.isfile(path):
+            raise ValueError("File not found: %s" % path)
+
+        line_iterator = open(path)
+
+    return line_iterator
+
+
+def _compute_tsv_header(obo_records):
+    """Compute .tsv file header as a list of strings containing all tags in the given obo_records
+
+    Args:
+        obo_records (iter): iterator over .obo records
+    """
+    all_tags = set()
+    for record in obo_records:
+        for tag in record:
+            all_tags.add(tag)
+
+    header = ['id', 'name']
+    other_columns = sorted(list(all_tags - set(EXCLUDE_FROM_TSV) - set(header)))
+    header.extend(other_columns)
+
+    return header
+
+
+def write_tsv(obo_records_dict, output_stream, root_id=None, separator=", "):
+    """Write obo_records_dict to the given output_stream.
+
+    Args:
+        obo_records_dict (dict): data structure returned by parse_obo_format(..)
+        output_stream (file): output stream where to write the .tsv file
+        root_id (str): Only ontology terms that are either descendants of the
+            given id or have this id themselves are returned. For example, 'HP:0000118'.
+        separator (str): separator for concatenating multiple values in a single column
+    """
+
+    header = _compute_tsv_header(obo_records_dict.values())
+    output_stream.write("\t".join([RENAME_COLUMNS.get(column, column) for column in header]))
+    output_stream.write("\n")
+    for record in get_substree(obo_records_dict, root_id):
+        row = []
+        for tag in header:
+            value = record.get(tag)
+            if value is None:
+                row.append("")
+            elif isinstance(value, list):
+                row.append(separator.join(map(str, value)))
+            else:
+                row.append(str(value))
+        output_stream.write("\t".join(row))
+        output_stream.write("\n")
+
+
+def _confirm_id_is_valid(term_id, obo_records_dict, label="id"):
+    """Raises an exception if the given term id doesn't exist in the given obo_records_dict."""
+
+    if term_id not in obo_records_dict:
+        raise ValueError("%s '%s' not found in ontology" % (label, term_id))
+
+
+if __name__ == "__main__":
+    p = argparse.ArgumentParser(description="Parse an .obo file and write out a .tsv table")
```

Comment:
I know it's just a parser, but I still don't really like single letter variable (outside of lambdas)